[core][rdt] RDT NIXL perf summary doc (#63307)
ray-project/ray: doc/source/ray-core/direct-transport/nixl-rdt-performance-tuning.rst

> .. _nixl-rdt-performance-tuning:

Reviewer: This page has no `.. meta::` description. Worth adding while the page is new.

The style guide's [front matter rule](https://docs.ray.io/en/master/ray-contribute/writing-style.html#front-matter) is written for MyST `html_meta`, so it doesn't literally cover an `.rst` page. The practice does: 62 of the 90 `.rst` pages under `ray-core` carry a `.. meta::` description, and 251 of 347 repo-wide. `direct-transport` is one of the gaps at 0 of 2, so your siblings don't model it, but the rest of `ray-core` does.

I drew the description from your own section headings rather than inventing claims, and matched the formatting of `cluster/faq.rst` (no quotes around the value, unlike the MyST form).

```suggestion
.. meta::
   :description: Tune Ray Direct Transport on the NIXL backend: PyTorch storage block registration, pre-registering memory with register_nixl_memory, receiving into pre-allocated tensors with set_target_for_ref, and NIXL agent reuse.

.. _nixl-rdt-performance-tuning:
```

Trim or reword freely. If @stephanie-wang's reorganization changes what the page covers, the description should follow it rather than this wording.

*****************************************
Performance tuning the NIXL RDT backend
*****************************************

Ray Direct Transport (RDT) layers a number of optimizations and helper APIs on top of `NVIDIA NIXL <https://github.com/ai-dynamo/nixl>`__ to reduce the per-transfer and memory registration overheads of RDMA.
This page explains what those optimizations are, when they help, and how to opt in to the helper APIs for the workloads that benefit the most.

If you haven't used RDT with NIXL before, start with the :ref:`NIXL walkthrough <direct-transport>` first.

Memory registration
===================

> NIXL builds on RDMA one-sided protocols, which require memory regions to be registered with the NIC before they can participate in a transfer.
> Registration is expensive: it issues a syscall per region and pins the pages so they can't be evicted.
> RDT provides two complementary optimizations to keep this cost off the hot path.

Reviewer: The guide [reserves the colon for introducing a list or a code block](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure). Here it joins two clauses, and the second one is really the interesting half, so it earns its own sentence.

```suggestion
NIXL builds on RDMA one-sided protocols, which require memory regions to be registered with the NIC before they can participate in a transfer.
Registration is expensive. It issues a syscall per region and pins the pages so they can't be evicted.
RDT provides two complementary optimizations to keep this cost off the hot path.
```

PyTorch storage block registration
----------------------------------

PyTorch uses its own memory allocator, so when you allocate a large 2D tensor, for example, the rows usually share a single contiguous storage block.

> Without this optimization, NIXL would register each row view individually, which can be very slow when you have many views.

Reviewer: `very` is on the guide's [cut-the-qualifiers list](https://docs.ray.io/en/master/ray-contribute/writing-style.html#cut-qualifiers-and-filler). Your own table two screens down quantifies exactly how slow, which is far better than "very," so pointing at it does more work than the adverb did.

```suggestion
Without this optimization, NIXL would register each row view individually, which gets slow once you have many views.
```

RDT instead registers the underlying PyTorch storage block once, regardless of how many tensor views point into it.
The following snippet contrasts the two approaches:

.. code-block:: python

    # Assume view_0 ... view_n share the same PyTorch storage block.

>     tensors = [view_0, view_1, view_2, ...]

Reviewer 2: Could we include actual code that can be tested here? Like creating actual views from a single torch tensor and maybe comparing nixl_agent.register_memory vs `ray.put` time on all views?

>     # Old behavior: n registrations, one per view.
>     for tensor in tensors:
>         nixl_agent.register_memory([tensor])
> 
>     # New behavior: one registration for the shared storage block.
>     nixl_agent.register_memory(

Reviewer: `Old behavior` and `New behavior` are dated the moment this ships. The guide asks you to [avoid time-relative framing](https://docs.ray.io/en/master/ray-contribute/writing-style.html#avoid-time-relative-words) because it goes stale as Ray evolves, and a reader arriving in six months has no idea which release "new" is relative to. Naming what each branch does instead stays true indefinitely.

```suggestion
    # Per-view registration: n registrations, one per view.
    for tensor in tensors:
        nixl_agent.register_memory([tensor])

    # Storage block registration: one registration for the shared block.
    nixl_agent.register_memory(
```

        [
            (

>                 view_0.untyped_storage().data_ptr(),

Reviewer 2: It'd be good to find a way to not include this because I don't think users will necessarily know about the significance of "storage" and "data_ptr".

                view_0.untyped_storage().nbytes(),
                ...
            )
        ]
    )

> The following table shows the registration and deregistration time for a 2D tensor with ``X`` rows of size ``Y`` each,

Reviewer 2: I'm a bit confused by what this table ist rying to say.

- Is it showing NIXL results or RDT+NIXL?
- What's the connection between these results and storage vs view-based registration? Is each row here a view? (I think it is?)

It might be clearer to just have a plot showing NIXL registration time vs # of views.

measured against the underlying storage block:

.. list-table::
   :header-rows: 1
   :widths: 25 25 25

   * - Tensor shape (X, Y)
     - Registration time (ms)
     - Deregistration time (ms)
   * - X = 1, Y = 2 GB
     - 10.570
     - 0.015
   * - X = 1000, Y = 200 KB
     - 857.315
     - 220.120
   * - X = 1000, Y = 2 MB
     - 1361.092
     - 262.175
   * - X = 10000, Y = 200 KB
     - 10449.209
     - 3113.972

> Registration time scales roughly linearly with the number of tensor views.

Reviewer 2: Should include performance plot for perf breakdown vs # tensor views.

Larger memory regions also increase registration time, but much less aggressively than the number of regions does.
Also note that the first registration on a fresh NIXL agent typically takes 5-10x longer than subsequent registrations on the same agent,

> so warm up the agent ahead of time if you're sensitive to first-call latency.

Reviewer 2: We should show the RDT+NIXL registration/ray.put time vs # of views and compare it to the NIXL results.

> Registration time scales roughly linearly with the number of tensor views.
> Larger memory regions also increase registration time, but much less aggressively than the number of regions does.
> Also note that the first registration on a fresh NIXL agent typically takes 5-10x longer than subsequent registrations on the same agent,
> so warm up the agent ahead of time if you're sensitive to first-call latency.

Reviewer: `Also note that` is on the guide's [filler list](https://docs.ray.io/en/master/ray-contribute/writing-style.html#cut-qualifiers-and-filler), and the sentence is stronger with the fact first and the instruction after. `5-10x` is fine as written.

```suggestion
Registration time scales roughly linearly with the number of tensor views.
Larger memory regions also increase registration time, but much less aggressively than the number of regions does.
The first registration on a fresh NIXL agent typically takes 5-10x longer than subsequent registrations on the same agent.
Warm up the agent ahead of time if you're sensitive to first-call latency.
```

        tensor = ray.get(rdt_list[0])

This default keeps the API consistent with regular Ray object lifetimes, but it has two drawbacks:

* Memory registration can happen on the hot path of your script every time you pass refs around.
* You have to keep the ``ObjectRef`` alive yourself to avoid paying for re-registration on the next transfer.

To keep registrations off the hot path, use :func:`ray.experimental.register_nixl_memory` to pre-register a tensor.
Ray keeps the registration alive until you call :func:`ray.experimental.deregister_nixl_memory` on the same tensor.
Use this API when you know the tensors you want to transfer up front and their memory location doesn't change.
A common pattern is registering the weight buffers used for weight syncs between a trainer and an inference engine:

.. code-block:: python

    from ray.experimental import register_nixl_memory, deregister_nixl_memory

    @ray.remote(num_gpus=1, enable_tensor_transport=True)
    class Trainer:
        def __init__(self):
            self.weights = torch.randn(1024 * 1024, device="cuda")
            register_nixl_memory(self.weights)

        def __del__(self):

>             deregister_nixl_memory(self.weights)

Reviewer 2: Do we need the __del__ (won't it normally be called on actor exit?) It'd be good to also briefly explain the deregister call.

Avoid intermediate copies with ``set_target_for_ref``
-----------------------------------------------------

By default, :func:`ray.get` materializes the result into a fresh tensor.
In the trainer example above, an ``update_weights`` method written the obvious way ends up allocating a throwaway buffer on every call:

.. code-block:: python

    def update_weights(self, weights_ref):
        # Allocates a new tensor and registers it with NIXL,
        # even though self.weights is already registered.
        new_weights = ray.get(weights_ref[0])
        self.weights = new_weights

That intermediate ``new_weights`` tensor wastes memory,
and with NIXL it's even more expensive because :func:`ray.get` registers the new tensor with NIXL before the transfer.

:func:`ray.experimental.set_target_for_ref` tells Ray to receive directly into a tensor that you already own
(and, ideally, that you've already registered):

.. code-block:: python

    from ray.experimental import set_target_for_ref

    def update_weights(self, weights_ref):
        set_target_for_ref(weights_ref[0], [self.weights])
        ray.get(weights_ref[0])

> The main limitation is that ``set_target_for_ref`` only works in the ``ray.put`` / borrowing pattern shown above.

Reviewer 2: Let's avoid using jargon like borrowing. You can just say ``ray.get``

> The main limitation is that ``set_target_for_ref`` only works in the ``ray.put`` / borrowing pattern shown above.
> It can't be used when the ref is passed as a task argument because Ray would already have created a placeholder tensor during argument deserialization before ``update_weights`` runs.

Reviewer: **This is a vote on @stephanie-wang's open thread, not a new request.** She [asked you to drop "borrowing"](https://github.com/ray-project/ray/pull/63307#discussion_r3503293357) as jargon and just say `ray.get`, and I agree, so I'm folding two mechanical items into the same rewrite rather than filing them separately:

- `ray.put / borrowing` uses a slash in prose, where the guide asks for [words instead of symbols](https://docs.ray.io/en/master/ray-contribute/writing-style.html#use-words-for-symbols-in-prose). Taking her suggestion removes it anyway.
- `It can't be used when the ref is passed as a task argument` is passive twice over with Ray named as the actor in the very next clause.

If you'd rather settle it on her thread first, ignore this and I'll follow along there.

```suggestion
``set_target_for_ref`` has one limitation: it works only with the ``ray.put`` and ``ray.get`` pattern shown above.
It doesn't work when you pass the ref as a task argument, because Ray creates a placeholder tensor during argument deserialization, before ``update_weights`` runs.
```

Note I changed "would already have created" to the present tense, since this describes current behavior rather than a hypothetical.

> For best results, combine ``set_target_for_ref`` with ``register_nixl_memory`` to prevent unnecessary memory registrations across calls.

Reviewer 2: You can make this a "Tip" box.

> NIXL agent reuse

Reviewer 2: I'd consider reorganizing this page into one section about APIs that give the user control vs more info on optimizations we provide on top of NIXL. I think it is slightly confusing right now when reading this section and the one on storage registration whether the user is expected to do something to enable this or not.

================

The NIXL agent on the receiver side isn't stateless.
It maintains a registry of remote agents (one per sender) and tracks every memory descriptor those senders have registered.
Each transfer goes through descriptor validation (``rkey``, memory regions, and so on) against this registry.

> The original implementation created and destroyed the remote agent entry on every receive.
> For senders with many distinct memory registrations, the per-transfer cost of this was significant.

Reviewer: Same staleness problem as the code comments above: `The original implementation` and the past tense date this to a specific release boundary that the page never names. A reader who arrives after this is the only behavior they've ever known has to work out what "original" refers to.

Describing the cost as a property of not reusing agents keeps it accurate however long the page lives, and it sets up the version-tracking paragraph that follows just as well.

```suggestion
Rebuilding the remote agent entry on every receive is expensive.
For senders with many distinct memory registrations, that per-transfer cost dominates.
```

Flagging one thing I can't judge: I've written this as though the non-reuse path still exists, which the `RAY_NIXL_REMOTE_AGENT_CACHE_MAXSIZE=0` escape hatch below suggests it does. If reuse is unconditional now and this really is history, keep the past tense and ignore me.

To avoid that overhead, Ray version-tracks each remote agent in the receiver's registry:

* When the sender adds a new memory registration, Ray incrementally updates the remote agent's descriptor list without bumping the version.
* When the sender deregisters memory, Ray bumps the version and rebuilds the remote agent, because NIXL doesn't support incremental deregistration.

> You can tune the cache size with the ``RAY_NIXL_REMOTE_AGENT_CACHE_MAXSIZE`` environment variable.

Reviewer 2: One thing that's not really clear is when / why you would want to do this? What guidance can we give to the user on how to set it?

> You can tune the cache size with the ``RAY_NIXL_REMOTE_AGENT_CACHE_MAXSIZE`` environment variable.
> Set it to ``0`` to disable agent reuse entirely (mostly useful for benchmarking).
> When the cache is full, Ray evicts the least recently used remote agent.

Reviewer: Two things, one of them a genuine gap.

**State the default.** A reader who's been told they can tune a cache size needs to know what they're tuning away from. It's `1000` (`python/ray/_private/ray_constants.py:600`), and without it the only actionable value on the page is `0`.

**The parenthetical.** The guide asks you to [restructure rather than use parentheticals in prose](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure). It's a small one, and folding it in reads fine.

```suggestion
You can tune the cache size with the ``RAY_NIXL_REMOTE_AGENT_CACHE_MAXSIZE`` environment variable, which defaults to ``1000``.
Set it to ``0`` to disable agent reuse entirely, which is mainly useful for benchmarking.
When the cache is full, Ray evicts the least recently used remote agent.
```

Double-check the default against your own reading of the constant before taking this. I read it off `ray_constants.py` at this head rather than from the PR's prose, but you know whether anything else overrides it.

The following table shows the cost of adding and removing a remote agent on the receiver side,
for a sender that has registered a 2D tensor with ``X`` rows of size ``Y`` each, with and without agent reuse:

.. list-table::
   :header-rows: 1
   :widths: 20 15 20 20

   * - Tensor shape (X, Y)
     - Agent reuse
     - Add remote agent (ms)
     - Remove remote agent (ms)
   * - X = 1, Y = 2 GB
     - false
     - 0.457
     - 0.020
   * - X = 1, Y = 2 GB
     - true
     - 0.001
     - 0.000
   * - X = 1000, Y = 2 MB
     - false
     - 1.469
     - 0.312
   * - X = 1000, Y = 2 MB
     - true
     - 0.279
     - 0.000
   * - X = 10000, Y = 200 KB
     - false
     - 15.658
     - 3.743
   * - X = 10000, Y = 200 KB
     - true
     - 3.128
     - 0.000

> The cost of initializing and tearing down a remote agent scales with the number of memory descriptors the sender has registered.
> Combining agent reuse with the storage block registration optimization collapses the descriptor count,
> which together bring all of the scenarios in the table down to roughly the cost of the ``X = 1000, Y = 2 MB`` row with ``agent_reuse = true``.

Reviewer: The second sentence doesn't parse: `Combining ... collapses the descriptor count, which together bring ...` has a singular subject with a plural verb, and `together` has nothing left to refer back to once `Combining` became the subject.

```suggestion
The cost of initializing and tearing down a remote agent scales with the number of memory descriptors the sender has registered.
Agent reuse and storage block registration together collapse the descriptor count.
Combining them brings every scenario in the table down to roughly the cost of the ``X = 1000, Y = 2 MB`` row with agent reuse enabled.
```

I also wrote "with agent reuse enabled" instead of ``agent_reuse = true``. `agent_reuse` isn't a setting a reader can find: it appears in the tree only as a test name, and the knob you documented above is `RAY_NIXL_REMOTE_AGENT_CACHE_MAXSIZE`. As written it reads like a config flag they should be able to look up. If it's meant purely as a pointer to the table's own column, the phrasing still costs you nothing.

Summary
=======

> For workloads that re-use the same tensors across many transfers, such as RL weight syncs:
> 
> * Re-use the same tensors and call :func:`ray.experimental.register_nixl_memory` once at startup, so memory registration stays off the hot path.
> * Pre-allocate receive buffers and bind them with :func:`ray.experimental.set_target_for_ref` to skip the intermediate copy on :func:`ray.get`.
> * Pass whole tensors, or views into a shared storage block, rather than many small standalone tensors, so RDT registers a single underlying storage block instead of many descriptors.

Reviewer: Three small things in the closing section:

- **`re-use` versus `reuse`.** The page says "NIXL agent reuse" and "with and without agent reuse" everywhere else, then switches to `re-use` twice here. Pick one; `reuse` is the standard form and matches your own heading.
- **The lead-in is a fragment.** The guide asks for [list lead-ins written as complete sentences](https://docs.ray.io/en/master/ray-contribute/writing-style.html#lists-and-punctuation) ending in a colon.
- **The first bullet repeats the lead-in.** Once the lead-in says "workloads that reuse the same tensors," the bullet's "Reuse the same tensors and" is redundant, and the real instruction is the `register_nixl_memory` call.

The file also needs a trailing newline.

```suggestion
For workloads that reuse the same tensors across many transfers, such as RL weight syncs, do the following:

* Call :func:`ray.experimental.register_nixl_memory` once at startup, so memory registration stays off the hot path.
* Pre-allocate receive buffers and bind them with :func:`ray.experimental.set_target_for_ref` to skip the intermediate copy on :func:`ray.get`.
* Pass whole tensors, or views into a shared storage block, rather than many small standalone tensors, so RDT registers a single underlying storage block instead of many descriptors.
```
